src/boot/efi/boot.c

```diff
@@ -1871,6 +1872,101 @@ static bool is_sd_boot(EFI_FILE *root_dir, const char16_t *loader_path) {
+        GUID_DEF(0x64074afe, 0x340a, 0x4be6, 0x94, 0xba, 0x91, 0xb5, 0xb4, 0xd0, 0xf7, 0x1e)
+#elif defined(__riscv)
+#  define EDK2_FIRMWARE_VOLUME_GUID \
+        GUID_DEF(0x27a72e80, 0x3118, 0x4c0c, 0x86, 0x73, 0xaa, 0x5b, 0x4e, 0xfa, 0x96, 0x13)
+#endif
+
+#ifdef EDK2_FIRMWARE_VOLUME_GUID
+static struct {
+        VENDOR_DEVICE_PATH fv;
+        VENDOR_DEVICE_PATH ff;
+        EFI_DEVICE_PATH end;
+} _packed_ edk2_firmware_uefi_shell_dp = {
+        .fv = {
+                .Header = {
+                        .Type = MEDIA_FILEPATH_DP,
+                        .SubType = MEDIA_PIWG_FW_VOL_DP,
+                        .Length = sizeof(edk2_firmware_uefi_shell_dp.fv),
+                },
+                .Guid = EDK2_FIRMWARE_VOLUME_GUID,
+        },
+        .ff = {
+                .Header = {
+                        .Type = MEDIA_FILEPATH_DP,
+                        .SubType = MEDIA_PIWG_FW_FILE_DP,
+                        .Length = sizeof(edk2_firmware_uefi_shell_dp.ff),
+                },
+                .Guid = UEFI_SHELL_GUID,
+        },
+        .end = DEVICE_PATH_END_NODE,
+};
+
+static EFI_STATUS launch_edk2_firmware_uefi_shell(const ConfigEntry *entry) {
+        EFI_STATUS err;
+
+        _cleanup_(unload_imagep) EFI_HANDLE shell = NULL;
+        err = BS->LoadImage(false, IMG, (EFI_DEVICE_PATH *) &edk2_firmware_uefi_shell_dp, NULL, 0, &shell);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error loading EDK2 firmware UEFI shell: %m");
+
+        err = BS->StartImage(shell, NULL, NULL);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error starting EDK2 firmware UEFI shell: %m");
+
+        return EFI_SUCCESS;
+}
+
+static void config_entry_add_edk2_firmware_uefi_shell(Config *config) {
+        /* Discovers the UEFI shell built into the EDK2 firmware volume. It can be loaded like any
+         * other UEFI application by using it's device path. Instead of trying to find the firmware
+         * volume and seeing if it has the shell file on it, we just ask LoadImage to do the job
+         * for us: if loading is successful we know it's available.
+         *
+         * On x86 the full device path for the internal UEFI shell is:
+         * Fv(7CB8BDC9-F8EB-4F34-AAEA-3EE4AF6516A1)/FvFile(7C04A583-9E3E-4F1C-AD65-E05268D0B4D1) */
+
+        assert(config);
+
+        if (!config->auto_entries || !config->auto_firmware)
+                return;
+
+        /* Let's make sure we're only doing this on EDK2 firmware inside a VM. Note that real
+         * firmware uses their own GUID for the firmware volume and wouldn't find this device path
+         * anyway. Nor does real firmware contain an internal shell anyway as it's quite large. */
+        if (!in_hypervisor())
+                return;
+
+        _cleanup_(unload_imagep) EFI_HANDLE shell = NULL;
+        if (BS->LoadImage(false, IMG, (EFI_DEVICE_PATH *) &edk2_firmware_uefi_shell_dp, NULL, 0, &shell) !=
```

Comment:
could also make edk2_firmware_uefi_shell_dp() a union, of an EFI_DEVICE_PATH adn the struct it currently is, and avoid the ugly cast